File: 094.py
#! /usr/bin/python
# Project Euler problem 94 ported from Mathematica code

# Timing Notes: It takes 21 seconds to go through the first ten million!
# We need to go up to 333 million

# This function works find!  Just takes too long.  Try it in C++.  Test C++ against this.

# Note that the factor (s-lst[0])*(s-lst[1]) is always a perfect square since these are 
# equal.  

# We just need to know when s*(s-lst[2]) is a perfect square.

from math import *
from sympy import *
import time

def Ara(lst): 
        s = sum(lst)/2
        return sqrt(s*(s-lst[0])*(s-lst[1])*(s-lst[2]))

mxi = 2e5
tol = 1e-8
mxp = 1e9
cnt = 0
prmtot = 0

t1 = time.clock()

#for i in xrange(2,int(mxi+1)):
for k in xrange(75000, int((mxi+1)/2.0)):
    i = 2*k-1
    tp1 = Ara([i,i,i-1])
    tp2 = Ara([i,i,i+1])
 #   t1err = abs(tp1 - round(tp1))
 #   t2err = abs(tp2 - round(tp2))

    if (tp1.is_integer is True and (3*i - 1) < mxp):
        cnt += 1
        prmtot += 3*i - 1
        print ((i,i,i-1), 3*i-1)
    if (tp2.is_integer is True and (3*i + 1) < mxp):
        cnt += 1
        prmtot += 3*i + 1
        print ((i,i,i+1), 3*i+1)

t2 = time.clock()

print "Count is: %r. Perm Tot is %r." %(cnt, prmtot)        
print "Total time is %r seconds." % (t2-t1)

File: 094.cpp
// Project Euler Problem 94 solution 

//We have an error in abs.  Find the proper usage of abs() in the math.h library!!! 

//Compute with: g++ 094.cpp -lgmp

#include <iostream>
#include <math.h>
#include <time.h>
#include <stdlib.h>

long double Ara(long int a, long int b, long int c)
{

    long double s = (a+b+c)/2.0;
    //    return  sqrt(s*(s-a)*(s-b)*(s-c));
    return sqrt(s*(s-c)); //Note that we only need this to be a perfect square
}

int main()
{
    long int mxi = 34*1e7;
    long double tol = 1e-8;
    long int mxp = 1e9;
    long int cnt = 0;
    long int prmtot = 0;

    clock_t start = clock();

    long double tp1 = 0;
    long double tp2 = 0;
    long double t1err = 0;
    long double t2err = 0;

    for(long int i = 2; i <= mxi; i++)
    {
        tp1 = Ara(i,i,i-1);
        tp2 = Ara(i,i,i+1);
        t1err = fabs(tp1 - round(tp1));
        t2err = fabs(tp2 - round(tp2));

        if((t1err < tol) && (3*i - 1) < mxp)
        {
            cnt++;
            prmtot += 3*i-1;
            std::cout << i << ",";
        }
        if((t2err < tol) && (3*i + 1) < mxp)
        {
            cnt++;
            prmtot += 3*i + 1;
            std::cout << i << ",";
        }
    }

    clock_t end = clock();
    float diff ((float)end - (float)start);

    std::cout << "Total time is: " << diff/CLOCKS_PER_SEC << std::endl;
    std::cout << "Count: " << cnt << " Perm Total: " << prmtot << std::endl;
    std::cout << tp1 << tp2 << t1err << t2err << std::endl;
    return 0;
}   

